compiler/rustc_const_eval/src/const_eval/valtrees.rs

```diff
@@ -0,0 +1,370 @@
+    ecx: &mut CompileTimeEvalContext<'tcx, 'tcx>,
+    ty: Ty<'tcx>,
+) -> MPlaceTy<'tcx> {
+    let tcx = ecx.tcx;
+    let param_env = ecx.param_env;
+    let layout = tcx.layout_of(param_env.and(ty)).unwrap();
+    debug!(?layout);
+
+    ecx.allocate(layout, MemoryKind::Stack).unwrap()
+}
+
+#[instrument(skip(ecx), level = "debug")]
+fn create_pointee_place<'tcx>(
+    ecx: &mut CompileTimeEvalContext<'tcx, 'tcx>,
+    ty: Ty<'tcx>,
+    valtree: ty::ValTree<'tcx>,
+) -> MPlaceTy<'tcx> {
+    let tcx = ecx.tcx.tcx;
+
+    match ty.kind() {
+        ty::Slice(_) | ty::Str => {
+            let slice_ty = match ty.kind() {
+                ty::Slice(slice_ty) => *slice_ty,
+                ty::Str => tcx.mk_ty(ty::Uint(ty::UintTy::U8)),
+                _ => bug!("expected ty::Slice | ty::Str"),
+            };
+
+            // Create a place for the underlying array
+            let len = valtree.unwrap_branch().len() as u64;
+            let arr_ty = tcx.mk_array(slice_ty, len as u64);
+            let place = create_mplace_from_layout(ecx, arr_ty);
+            debug!(?place);
+
+            place
+        }
+        _ => create_mplace_from_layout(ecx, ty),
+    }
+}
+
+/// Converts a `ValTree` to a `ConstValue`, which is needed after mir
+/// construction has finished.
+#[instrument(skip(tcx), level = "debug")]
+pub fn valtree_to_const_value<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env_ty: ty::ParamEnvAnd<'tcx, Ty<'tcx>>,
+    valtree: ty::ValTree<'tcx>,
+) -> ConstValue<'tcx> {
+    // Basic idea: We directly construct `Scalar` values from trivial `ValTree`s
+    // (those for constants with type bool, int, uint, float or char).
+    // For all other types we create an `MPlace` and fill that by walking
+    // the `ValTree` and using `place_projection` and `place_field` to
+    // create inner `MPlace`s which are filled recursively.
+    // FIXME Does this need an example?
+
+    let (param_env, ty) = param_env_ty.into_parts();
+    let mut ecx = mk_eval_cx(tcx, DUMMY_SP, param_env, false);
+
+    match ty.kind() {
+        ty::Bool | ty::Int(_) | ty::Uint(_) | ty::Float(_) | ty::Char => match valtree {
+            ty::ValTree::Leaf(scalar_int) => ConstValue::Scalar(Scalar::Int(scalar_int)),
+            ty::ValTree::Branch(_) => bug!(
+                "ValTrees for Bool, Int, Uint, Float or Char should have the form ValTree::Leaf"
+            ),
+        },
+        ty::Ref(_, inner_ty, _) => {
+            // create a place for the pointee
+            let mut pointee_place = create_pointee_place(&mut ecx, *inner_ty, valtree);
+            debug!(?pointee_place);
+
+            // insert elements of valtree into `place`
+            fill_place_recursively(&mut ecx, &mut pointee_place, valtree);
+            dump_place(&ecx, pointee_place.into());
+            intern_const_alloc_recursive(&mut ecx, InternKind::Constant, &pointee_place).unwrap();
+
+            let ref_place = pointee_place.to_ref(&tcx);
```

Comment:
Except for this line and the following one, this match arm is equal to the next one. Please merge them and run these two lines conditionally by matching on the type within the arm